packages/langium/src/workspace/documents.ts

```diff
@@ -198,33 +211,97 @@ export class DefaultLangiumDocumentFactory implements LangiumDocumentFactory {
-        let textDoc: TextDocument | undefined = undefined;
+    protected createLangiumDocument<T extends AstNode = AstNode>(parseResult: ParseResult<T>, uri: URI, textDocument?: TextDocument, text?: string): LangiumDocument<T> {
+        let document: LangiumDocument<T>;
+        if (textDocument) {
+            document = {
+                parseResult,
+                uri: uri,
+                state: DocumentState.Parsed,
+                references: [],
+                textDocument
+            };
+        } else {
+            const documentTemplate = this.createTextDocumentTemplate(uri, text);
+            const textDocumentGetter = this.getTextDocumentGetter(documentTemplate);
+            document = {
+                parseResult,
+                uri: uri,
+                state: DocumentState.Parsed,
+                references: [],
+                get textDocument() {
+                    return textDocumentGetter();
+                }
+            };
+        }
+        (parseResult.value as Mutable<AstNode>).$document = document;
+        return document;
+    }
+
+    update<T extends AstNode = AstNode>(document: Mutable<LangiumDocument<T>>): LangiumDocument<T> {
+        const textDocument = this.textDocuments.get(document.uri.toString());
+        const text = textDocument ? textDocument.getText() : this.getContent(document.uri);
+
+        if (textDocument) {
+            Object.defineProperty(
+                document, 'textDocument', {
+                    value: textDocument
+                }
+            );
+        } else {
+            const documentTemplate = this.createTextDocumentTemplate(document.uri, text);
+            const textDocumentGetter = this.getTextDocumentGetter(documentTemplate);
+            Object.defineProperty(
+                document, 'textDocument', {
+                    get: textDocumentGetter
+                }
+            );
+        }
+
+        document.parseResult = this.parse(document.uri, text);
+        (document.parseResult.value as Mutable<AstNode>).$document = document;
+        document.state = DocumentState.Parsed;
+        return document;
+    }
+
+    /** Counterpart of {@link DefaultLangiumDocuments.getContent}. */
+    protected getContent(uri: URI): string {
+        return this.fileSystemProvider.readFileSync(uri);
+    }
+
+    protected parse<T extends AstNode>(uri: URI, text: string): ParseResult<T> {
+        const services = this.serviceRegistry.getServices(uri);
+        return services.parser.LangiumParser.parse<T>(text);
+    }
+
+    protected createTextDocumentTemplate(uri: URI, text?: string): TextDocumentTemplate {
+        const languageId = this.serviceRegistry.getServices(uri).LanguageMetaData.languageId;
+        return {
+            $template: true,
+            languageId,
+            uri,
+            text
+        };
+    }
+
+    protected getTextDocumentGetter(template: TextDocumentTemplate): () => TextDocument {
```

Comment:
Now to simplify this further, you can remove the `TextDocumentTemplate` type and pass the `uri` and `text` arguments directly to this method.